adcc/gradients/orbital_response.py

```diff
@@ -0,0 +1,340 @@
+            - einsum("jkia,ka->ij", hf.ooov, g1o.ov)
+            - einsum("jaib,ab->ij", hf.ovov, g1o.vv)
+        )
+        w.vv = gi_vv - einsum("ac,cb->ab", g1o.vv, hf.fvv)
+    return evaluate(w)
+
+
+class OrbitalResponseMatrix:
+    def __init__(self, hf):
+        self.hf = hf
+
+    @property
+    def shape(self):
+        no1 = self.hf.mospaces.n_orbs(b.o)
+        if self.hf.has_core_occupied_space:
+            no1 += self.hf.mospaces.n_orbs(b.c)
+        nv1 = self.hf.mospaces.n_orbs(b.v)
+        size = no1 * nv1
+        return (size, size)
+
+    def __matmul__(self, lam):
+        ret_ov = (
+            + einsum("ab,ib->ia", self.hf.fvv, lam.ov)
+            - einsum("ij,ja->ia", self.hf.foo, lam.ov)
+            + einsum("ijab,jb->ia", self.hf.oovv, lam.ov)
+            - einsum("ibja,jb->ia", self.hf.ovov, lam.ov)
+        )
+        if self.hf.has_core_occupied_space:
+            ret_ov += (
+                + einsum("iJab,Jb->ia", self.hf.ocvv, lam.cv)
+                - einsum("ibJa,Jb->ia", self.hf.ovcv, lam.cv)
+            )
+            ret_cv = (
+                + einsum("ab,Ib->Ia", self.hf.fvv, lam.cv)
+                - einsum("IJ,Ja->Ia", self.hf.fcc, lam.cv)
+                + einsum("IJab,Jb->Ia", self.hf.ccvv, lam.cv)
+                - einsum("IbJa,Jb->Ia", self.hf.cvcv, lam.cv)
+                + einsum("Ijab,jb->Ia", self.hf.covv, lam.ov)
+                - einsum("Ibja,jb->Ia", self.hf.cvov, lam.ov)
+            )
+            ret = AmplitudeVector(cv=ret_cv, ov=ret_ov)
+        else:
+            ret = AmplitudeVector(ov=ret_ov)
+        return evaluate(ret)
+
+
+class OrbitalResponsePinv:
+    def __init__(self, hf):
+        self.hf = hf
+        # Terms common to adc and cvs-adc
+        fo = hf.fock(b.oo).diagonal()
+        fv = hf.fock(b.vv).diagonal()
+        fov = direct_sum("-i+a->ia", fo, fv)
+
+        if hf.has_core_occupied_space:
+            fc = hf.fock(b.cc).diagonal()
+            fcv = direct_sum("-I+a->Ia", fc, fv)
+            self.df = AmplitudeVector(cv=fcv, ov=fov)
+        else:
+            self.df = AmplitudeVector(ov=fov)
+        self.df.evaluate()
+
+    @property
+    def shape(self):
+        no1 = self.hf.mospaces.n_orbs(b.o)
+        if self.hf.has_core_occupied_space:
+            no1 += self.hf.mospaces.n_orbs(b.c)
+        nv1 = self.hf.mospaces.n_orbs(b.v)
+        size = no1 * nv1
+        return (size, size)
+
+    def __matmul__(self, invec):
+        return invec / self.df
+
+
+def orbital_response(hf, rhs):
+    """
+    Solves the orbital response equations
+    for a given reference state and right-hand side
+    """
+    # TODO: pass solver arguments
+    A = OrbitalResponseMatrix(hf)
+    Pinv = OrbitalResponsePinv(hf)
+    x0 = (Pinv @ rhs).evaluate()
+    lam = conjugate_gradient(A, rhs=rhs, x0=x0, Pinv=Pinv,
+                             explicit_symmetrisation=None,
+                             callback=default_print)
```

Comment:
Which tolerance is used and why? Should for sure be related to the convergence used for the ADC excitation vectors ... or maybe aim for relative residual decrease?

Reply:
I think it should rather be related to the SCF convergence criterion, because we compute the response of the HF ground state (kind of 😄). In addition, MP states don't have an additional convergence criterion, so I will make this depend on the SCF conv tol. 👍